File: 2023/day18/go/solution.go
// Day 18: Lavaduct Lagoon - Polygon area with Shoelace formula and Pick's theorem.
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Instruction struct {
	direction string
	distance  int
	color     string
}

type Point struct {
	r, c int
}

func parseInput(filename string) ([]Instruction, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var instructions []Instruction
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		parts := strings.Fields(line)
		direction := parts[0]
		distance, _ := strconv.Atoi(parts[1])
		// Extract color code: "(#70c710)" -> "70c710"
		color := parts[2][2 : len(parts[2])-1]
		instructions = append(instructions, Instruction{direction, distance, color})
	}
	return instructions, scanner.Err()
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// calculateArea computes total area using Shoelace formula and Pick's theorem.
// Shoelace gives twice the signed area of the polygon.
// Pick's theorem: A = i + b/2 - 1, where i = interior points, b = boundary points
// We want: Total = i + b = A + b/2 + 1
func calculateArea(vertices []Point, perimeter int) int {
	n := len(vertices)
	area := 0
	for i := 0; i < n; i++ {
		j := (i + 1) % n
		area += vertices[i].r * vertices[j].c
		area -= vertices[j].r * vertices[i].c
	}
	area = abs(area) / 2

	// Total points = interior + boundary
	// From Pick's theorem: interior = area - boundary/2 + 1
	// Total = interior + boundary = area + boundary/2 + 1
	return area + perimeter/2 + 1
}

func part1(instructions []Instruction) int {
	directionMap := map[string]Point{
		"R": {0, 1},
		"D": {1, 0},
		"L": {0, -1},
		"U": {-1, 0},
	}

	vertices := []Point{{0, 0}}
	perimeter := 0
	r, c := 0, 0

	for _, inst := range instructions {
		delta := directionMap[inst.direction]
		r += delta.r * inst.distance
		c += delta.c * inst.distance
		vertices = append(vertices, Point{r, c})
		perimeter += inst.distance
	}

	return calculateArea(vertices, perimeter)
}

func part2(instructions []Instruction) int {
	// Last digit of hex: 0=R, 1=D, 2=L, 3=U
	// First 5 digits: distance in hex
	directionMap := map[byte]Point{
		'0': {0, 1},   // R
		'1': {1, 0},   // D
		'2': {0, -1},  // L
		'3': {-1, 0},  // U
	}

	vertices := []Point{{0, 0}}
	perimeter := 0
	r, c := 0, 0

	for _, inst := range instructions {
		// Parse distance from first 5 hex digits
		distance, _ := strconv.ParseInt(inst.color[:5], 16, 64)
		// Get direction from last hex digit
		dirChar := inst.color[5]
		delta := directionMap[dirChar]
		r += delta.r * int(distance)
		c += delta.c * int(distance)
		vertices = append(vertices, Point{r, c})
		perimeter += int(distance)
	}

	return calculateArea(vertices, perimeter)
}

func main() {
	instructions, err := parseInput("../input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Part 1: %d\n", part1(instructions))
	fmt.Printf("Part 2: %d\n", part2(instructions))
}

File: 2023/day18/rust/solution.rs
// Day 18: Lavaduct Lagoon - Polygon area with Shoelace formula and Pick's theorem

use std::fs;
use std::path::Path;

/// Parse a single instruction line into (direction, distance, hex_color)
fn parse_line(line: &str) -> (char, i64, &str) {
    let parts: Vec<&str> = line.split_whitespace().collect();
    let direction = parts[0].chars().next().unwrap();
    let distance = parts[1].parse::<i64>().unwrap();
    // Extract hex color: "(#70c710)" -> "70c710"
    let color = &parts[2][2..parts[2].len() - 1];
    (direction, distance, color)
}

/// Calculate total area using Shoelace formula and Pick's theorem.
/// Shoelace gives twice the signed area of the polygon.
/// Pick's theorem: A = i + b/2 - 1, where i = interior points, b = boundary points
/// We want: Total = i + b = A + b/2 + 1
fn calculate_area(vertices: &[(i64, i64)], perimeter: i64) -> i64 {
    let n = vertices.len();
    let mut area: i64 = 0;

    for i in 0..n {
        let j = (i + 1) % n;
        area += vertices[i].0 * vertices[j].1;
        area -= vertices[j].0 * vertices[i].1;
    }
    area = area.abs() / 2;

    // Total points = interior + boundary = area + boundary/2 + 1
    area + perimeter / 2 + 1
}

/// Part 1: Follow the dig plan directions
fn part1(lines: &[&str]) -> i64 {
    let mut vertices = vec![(0i64, 0i64)];
    let mut perimeter = 0i64;
    let mut r = 0i64;
    let mut c = 0i64;

    for line in lines {
        let (direction, distance, _) = parse_line(line);
        let (dr, dc) = match direction {
            'R' => (0, 1),
            'D' => (1, 0),
            'L' => (0, -1),
            'U' => (-1, 0),
            _ => panic!("Unknown direction: {}", direction),
        };

        r += dr * distance;
        c += dc * distance;
        vertices.push((r, c));
        perimeter += distance;
    }

    calculate_area(&vertices, perimeter)
}

/// Part 2: Decode instructions from hex color codes
/// Last digit of hex: 0=R, 1=D, 2=L, 3=U
/// First 5 digits: distance in hex
fn part2(lines: &[&str]) -> i64 {
    let mut vertices = vec![(0i64, 0i64)];
    let mut perimeter = 0i64;
    let mut r = 0i64;
    let mut c = 0i64;

    for line in lines {
        let (_, _, color) = parse_line(line);

        // First 5 hex digits are the distance
        let distance = i64::from_str_radix(&color[..5], 16).unwrap();
        // Last digit is the direction
        let dir_code = color.chars().nth(5).unwrap();

        let (dr, dc) = match dir_code {
            '0' => (0, 1),   // R
            '1' => (1, 0),   // D
            '2' => (0, -1),  // L
            '3' => (-1, 0),  // U
            _ => panic!("Unknown direction code: {}", dir_code),
        };

        r += dr * distance;
        c += dc * distance;
        vertices.push((r, c));
        perimeter += distance;
    }

    calculate_area(&vertices, perimeter)
}

fn main() {
    let input_path = Path::new(file!()).parent().unwrap().join("../input.txt");
    let content = fs::read_to_string(&input_path)
        .unwrap_or_else(|_| fs::read_to_string("../input.txt").expect("Failed to read input.txt"));

    let lines: Vec<&str> = content.trim().lines().collect();

    println!("Part 1: {}", part1(&lines));
    println!("Part 2: {}", part2(&lines));
}
